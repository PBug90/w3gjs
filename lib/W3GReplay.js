const fs = require('fs')
const zlib = require('zlib')
const { Parser } = require('binary-parser')
const { ActionBlockList } = require('../parsers/actions')
const { ReplayHeader,
  EncodedMapMetaString,
  GameMetaData} = require('../parsers/header')

const { GameDataParser } = require('../parsers/gamedata')

const GameDataParserComposed = new Parser()
  .nest('meta', {type: GameMetaData})
  .nest('blocks', {type: GameDataParser})

const Player = require('./Player')
const crypto = require('crypto')

function W3GReplay () {
  return this
}

W3GReplay.prototype.parse = function (filepath) {
  console.time('parse')
  this.buffer = fs.readFileSync(filepath)
  this.buffer = this.buffer.slice(this.buffer.indexOf('Warcraft III recorded game'))

  this.header = ReplayHeader.parse(this.buffer)
  const decompressed = []
  this.header.blocks.forEach((block) => {
    if (block.blockSize > 0 && block.blockDecompressedSize === 8192) {
      try {
        const r = zlib.inflateSync(block.compressed, { finishFlush: zlib.constants.Z_SYNC_FLUSH })
        if (r.byteLength > 0 && block.compressed.byteLength > 0) {
          decompressed.push(r)
        }
      } catch (ex) {
        console.log(ex)
      }
    }
  })
  this.decompressed = Buffer.concat(decompressed)
  this.gameMetaDataDecoded = GameDataParserComposed.parse(this.decompressed)

  this.decodedMetaStringBuffer = this.decodeGameMetaString(this.gameMetaDataDecoded.meta.encodedString)
  this.meta = Object.assign({}, this.gameMetaDataDecoded, EncodedMapMetaString.parse(this.decodedMetaStringBuffer))

  this.slots = this.gameMetaDataDecoded.meta.playerSlotRecords
  this.playerList = [this.gameMetaDataDecoded.meta.player, ...this.gameMetaDataDecoded.meta.playerList]

  this.playerActionTrackInterval = 60000 // apm tracking interval in milliseconds

  this.createPlayerList()
  this.processGameDataBlocks()
  this.cleanup()

  return this.finalize()
}

W3GReplay.prototype.createPlayerList = function () {
  const tempPlayers = {}
  this.teams = {}
  this.players = {}

  this.playerList.forEach(player => {
    tempPlayers[player.playerId] = player
  })

  this.slots.forEach(slot => {
    if (slot.slotStatus > 1) {
      this.players[slot.playerId] = { ...this.players[slot.playerId], ...slot }
      this.teams[slot.teamId] = this.teams[slot.teamId] || []
      this.teams[slot.teamId].push(slot.playerId)
      this.players[slot.playerId] =
        new Player(slot.playerId, tempPlayers[slot.playerId] ? tempPlayers[slot.playerId].playerName : 'Computer', slot.teamId, slot.color, slot.raceFlag)
    }
  })
}

W3GReplay.prototype.processGameDataBlocks = function () {
  this.chatlog = []
  this.leaveEvents = []
  this.w3mmd = []

  this.totalTimeTracker = 0
  this.timeSegmentTracker = 0
  this.gameMetaDataDecoded.blocks.forEach((block) => {
    switch (block.type) {
      case 31:
      case 30:
        this.totalTimeTracker += block.timeIncrement
        this.timeSegmentTracker += block.timeIncrement
        if (this.timeSegmentTracker > this.playerActionTrackInterval) {
          Object.values(this.players).forEach((p) => p.newActionTrackingSegment())
          this.timeSegmentTracker = 0
        }
        this.processTimeSlot(block)
        break
      case 32:
        block.timeMS = this.totalTimeTracker
        this.chatlog.push(block)
        break
      case 23:
        this.leaveEvents.push(block)
        break
    }
  })

  this.chatlog = this.chatlog.map(elem => ({...elem, playerName: this.players[elem.playerId].name}))
  delete this.temporaryAPMTracker
}

W3GReplay.prototype.processTimeSlot = function (timeSlotBlock) {
  timeSlotBlock.actions.forEach(actionBlock => {
    this.processCommandDataBlock(actionBlock)
  })
}

W3GReplay.prototype.processCommandDataBlock = function (actionBlock) {
  const currentPlayer = this.players[actionBlock.playerId]
  let lastActionWasDeselect = false
  currentPlayer.currentTimePlayed = this.totalTimeTracker
  try {
    ActionBlockList.parse(actionBlock.actions).forEach(action => {
      switch (action.actionId) {
        case 0x10:
          currentPlayer.handle0x10(action.itemId, this.totalTimeTracker)
          break
        case 0x11:
          currentPlayer.handle0x11(action.itemId, this.totalTimeTracker)
          break
        case 0x12:
          currentPlayer.handle0x12(action.itemId, this.totalTimeTracker)
          break
        case 0x13:
          currentPlayer.handle0x13(action.itemId)
          break
        case 0x14:
          currentPlayer.handle0x14(action.itemId1)
          break
        case 0x16:
          if (action.selectMode === 0x02) {
            lastActionWasDeselect = true
            currentPlayer.handle0x16(action.selectMode, true)
          } else {
            if (lastActionWasDeselect === false) {
              currentPlayer.handle0x16(action.selectMode, true)
            }
            lastActionWasDeselect = false
          }
          break
        case 0x17:
        case 0x18:
        case 0x1C:
        case 0x1D:
        case 0x1E:
        case 0x61:
        case 0x65:
        case 0x66:
        case 0x67:
          currentPlayer.handleOther(action.actionId)
          break
        case 0x6b:
          this.w3mmd.push(action)
          break
      }
    })
  } catch (ex) {
    console.error(ex)
  }
}

W3GReplay.prototype.decodeGameMetaString = function (str) {
  let test = Buffer.from(str, 'hex')
  let decoded = Buffer.alloc(test.length)
  let mask = 0
  let dpos = 0

  for (let i = 0; i < test.length; i++) {
    if (i % 8 === 0) {
      mask = test[i]
    } else {
      if ((mask & (0x1 << (i % 8))) === 0) {
        decoded.writeUInt8(test[i] - 1, dpos++)
      } else {
        decoded.writeUInt8(test[i], dpos++)
      }
    }
  }
  return decoded
}

W3GReplay.prototype.isObserver = function (player) {
  return (player.teamid === 24 && this.header.version >= 29) || (player.teamid === 12 && this.header.version < 29)
}

W3GReplay.prototype.determineMatchup = function () {
  let teamRaces = {}
  Object.values(this.players).forEach((p) => {
    if (!this.isObserver(p)) {
      teamRaces[p.teamid] = teamRaces[p.teamid] || []
      teamRaces[p.teamid].push(p.raceDetected || p.race)
    }
  })
  this.gametype = Object.values(teamRaces).map(e => e.length).sort().join('on')
  this.matchup = Object.values(teamRaces).map(e => e.sort().join('')).sort().join('v')
}

W3GReplay.prototype.generateID = function () {
  let players = Object.values(this.players).filter((p) => this.isObserver(p) === false).sort((player1, player2) => {
    if (player1.id < player2.id) {
      return -1
    }
    return 1
  }).reduce((accumulator, player) => {
    accumulator += `${player.name}_${player.id}_${player.teamid}`
    return accumulator
  }, '')
  const idBase = this.meta.meta.randomSeed + players + this.meta.mapName
  this.id = crypto.createHash('sha256').update(idBase).digest('hex')
}

W3GReplay.prototype.cleanup = function () {
  this.determineMatchup()
  this.generateID()
  this.observers = []

  Object.values(this.players).forEach(p => {
    p.newActionTrackingSegment(this.playerActionTrackInterval)
    p.cleanup()
    if (this.isObserver(p)) {
      this.observers.push(p.name)
      delete this.players[p.id]
    }
  })

  if (this.header.version >= 29 && this.teams.hasOwnProperty('24')) {
    delete this.teams['24']
  } else if (this.teams.hasOwnProperty('12')) {
    delete this.teams['12']
  }
  delete this.slots
  delete this.playerList
  delete this.buffer
  delete this.decompressed
  delete this.gameMetaDataDecoded
  delete this.decodedMetaStringBuffer
  delete this.header.blocks
  delete this.apmTimeSeries
  delete this.meta.blocks
}

W3GReplay.prototype.finalize = function () {
  const settings = {
    referees: Boolean(this.meta.referees),
    fixedTeams: Boolean(this.meta.fixedTeams),
    fullSharedUnitControl: Boolean(this.meta.fullSharedUnitControl),
    alwaysVisible: Boolean(this.meta.alwaysVisible),
    hideTerrain: Boolean(this.meta.hideTerrain),
    mapExplored: Boolean(this.meta.mapExplored),
    teamsTogether: Boolean(this.meta.teamsTogether),
    randomHero: Boolean(this.meta.randomHero),
    randomRaces: Boolean(this.meta.randomRaces),
    speed: this.meta.speed
  }
  const root = {
    id: this.id,
    gamename: this.meta.meta.gameName,
    randomseed: this.meta.meta.randomSeed,
    startSpots: this.meta.meta.startSpotCount,
    observers: this.observers,
    players: Object.values(this.players).sort((player1, player2) => player2.teamid >= player1.teamid && player2.id > player1.id ? -1 : 1),
    matchup: this.matchup,
    creator: this.meta.creator,
    type: this.gametype,
    chat: [],
    apm: {
      trackingInterval: this.playerActionTrackInterval
    },
    map: {
      path: this.meta.mapName,
      file: this.meta.mapName.split('\\').pop(),
      checksum: this.meta.mapChecksum
    },
    version: `1.${this.header.version}`,
    duration: this.header.replayLengthMS,
    expansion: this.header.gameIdentifier === 'PX3W',
    settings
  }

  return root
}

module.exports = W3GReplay
