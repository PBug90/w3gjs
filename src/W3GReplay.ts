import { ActionBlockList } from './parsers/actions'
import Player from './Player'
import convert from './convert'
import ReplayParser from './ReplayParser'
import {
    GameMetaDataDecoded,
    SlotRecord,
    GameDataBlock,
    ActionBlock,
    TimeSlotBlock,
    CommandDataBlock,
    ParserOutput,
    PlayerChatMessageBlock
} from './types'
import { sortPlayers } from './sort'

// Cannot import node modules directly because error with rollup
// https://rollupjs.org/guide/en#error-name-is-not-exported-by-module-
const { createHash } = require('crypto')
const {
    performance
} = require('perf_hooks')

class W3GReplay extends ReplayParser {
    players: { [key: string]: Player }

    observers: string[]

    chatlog: any

    playerActionTracker: {[key: string]: any[]} = {}

    id = ''

    leaveEvents: any[]

    w3mmd: any[]

    slots: any[]

    teams: any[]

    meta: GameMetaDataDecoded

    playerList: any[]

    totalTimeTracker = 0

    timeSegmentTracker = 0

    playerActionTrackInterval = 60000

    gametype = ''

    matchup = ''

    parseStartTime: number

    constructor () {
        super()
        this.on('gamemetadata', (metaData: GameMetaDataDecoded) => this.handleMetaData(metaData))
        this.on('gamedatablock', (block: GameDataBlock) => this.processGameDataBlock(block))
        this.on('timeslotblock', (block: TimeSlotBlock) => this.handleTimeSlot(block))
    }

    // gamedatablock timeslotblock commandblock actionblock
    parse ($buffer: string | Buffer): ParserOutput {
        this.parseStartTime = performance.now()
        this.buffer = Buffer.from('')
        this.filename = ''
        this.id = ''
        this.chatlog = []
        this.leaveEvents = []
        this.w3mmd = []
        this.players = {}
        this.totalTimeTracker = 0
        this.timeSegmentTracker = 0
        this.playerActionTrackInterval = 60000

        super.parse($buffer)

        this.chatlog = this.chatlog.map((elem: PlayerChatMessageBlock) => {
            return ({ ...elem, player: this.players[elem.playerId].name })
        })

        this.generateID()
        this.determineMatchup()
        this.cleanup()

        return this.finalize()
    }

    handleMetaData (metaData: GameMetaDataDecoded) {
        this.slots = metaData.playerSlotRecords
        this.playerList = [metaData.player, ...metaData.playerList]
        this.meta = metaData
        const tempPlayers: {[key: string]: GameMetaDataDecoded['player'] } = {}
        this.teams = []
        this.players = {}

        this.playerList.forEach((player: GameMetaDataDecoded['player']): void => {
            tempPlayers[player.playerId] = player
        })

        this.slots.forEach((slot: SlotRecord) => {
            if (slot.slotStatus > 1) {
                this.teams[slot.teamId] = this.teams[slot.teamId] || []
                this.teams[slot.teamId].push(slot.playerId)

                this.players[slot.playerId] = new Player(slot.playerId, tempPlayers[slot.playerId]
                    ? tempPlayers[slot.playerId].playerName
                    : 'Computer', slot.teamId, slot.color, slot.raceFlag)
            }
        })
    }

    processGameDataBlock (block: GameDataBlock) {
        switch (block.type) {
            case 31:
            case 30:
                this.totalTimeTracker += block.timeIncrement
                this.timeSegmentTracker += block.timeIncrement
                if (this.timeSegmentTracker > this.playerActionTrackInterval) {
                    // @ts-ignore
                    Object.values(this.players).forEach(p => p.newActionTrackingSegment())
                    this.timeSegmentTracker = 0
                }
                break
            case 32:
                block.timeMS = this.totalTimeTracker
                this.chatlog.push(block)
                break
            case 23:
                this.leaveEvents.push(block)
                break
        }
    }

    handleTimeSlot (block: TimeSlotBlock): void {
        block.actions.forEach((commandBlock: CommandDataBlock): void => {
            this.processCommandDataBlock(commandBlock)
        })
    }

    processCommandDataBlock (block: CommandDataBlock) {
        const currentPlayer = this.players[block.playerId]
        currentPlayer.currentTimePlayed = this.totalTimeTracker
        currentPlayer._lastActionWasDeselect = false
        try {
            ActionBlockList.parse(block.actions).forEach((action: ActionBlock): void => {
                this.handleActionBlock(action, currentPlayer)
            })
        } catch (ex) {
            console.error(ex)
        }
    }

    handleActionBlock (action: ActionBlock, currentPlayer: Player) {
        this.playerActionTracker[currentPlayer.id] = this.playerActionTracker[currentPlayer.id] || []
        this.playerActionTracker[currentPlayer.id].push(action)

        if (action.itemId && (action.itemId.value === 'tert' || action.itemId.value === 'tret')) {
            currentPlayer.handleRetraining(this.totalTimeTracker)
        }
        switch (action.actionId) {
            case 0x10:
                currentPlayer.handle0x10(action.itemId, this.totalTimeTracker)
                break
            case 0x11:
                currentPlayer.handle0x11(action.itemId, this.totalTimeTracker)
                break
            case 0x12:
                currentPlayer.handle0x12(action.itemId)
                break
            case 0x13:
                currentPlayer.handle0x13(action.itemId)
                break
            case 0x14:
                currentPlayer.handle0x14(action.itemId1)
                break
            case 0x16:
                if (action.selectMode === 0x02) {
                    currentPlayer._lastActionWasDeselect = true
                    currentPlayer.handle0x16(action.selectMode, true)
                } else {
                    if (currentPlayer._lastActionWasDeselect === false) {
                        currentPlayer.handle0x16(action.selectMode, true)
                    }
                    currentPlayer._lastActionWasDeselect = false
                }
                break
            case 0x17:
            case 0x18:
            case 0x1C:
            case 0x1D:
            case 0x1E:
            case 0x61:
            case 0x65:
            case 0x66:
            case 0x67:
                currentPlayer.handleOther(action.actionId)
                break
            case 0x6b:
                this.w3mmd.push(action)
                break
        }
    }

    isObserver (player: Player): boolean {
        return (player.teamid === 24 && this.header.version >= 29) || (player.teamid === 12 && this.header.version < 29)
    }

    determineMatchup (): void {
        const teamRaces: { [key: string]: string[] } = {}
        Object.values(this.players).forEach(p => {
            if (!this.isObserver(p)) {
                teamRaces[p.teamid] = teamRaces[p.teamid] || []
                teamRaces[p.teamid].push(p.raceDetected || p.race)
            }
        })
        this.gametype = Object.values(teamRaces).map(e => e.length).sort().join('on')
        this.matchup = Object.values(teamRaces).map(e => e.sort().join('')).sort().join('v')
    }

    generateID (): void {
        const players = Object.values(this.players).filter((p) => this.isObserver(p) === false).sort((player1, player2) => {
            if (player1.id < player2.id) {
                return -1
            }
            return 1
        }).reduce((accumulator, player) => {
            accumulator += player.name
            return accumulator
        }, '')

        const idBase = this.meta.randomSeed + players + this.meta.mapName
        this.id = createHash('sha256').update(idBase).digest('hex')
    }

    cleanup (): void {
        this.observers = []

        Object.values(this.players).forEach(p => {
            p.newActionTrackingSegment(this.playerActionTrackInterval)
            p.cleanup()
            if (this.isObserver(p)) {
                this.observers.push(p.name)
                delete this.players[p.id]
            }
        })

        if (this.header.version >= 29 && Object.prototype.hasOwnProperty.call(this.teams, '24')) {
            delete this.teams[24]
        } else if (Object.prototype.hasOwnProperty.call(this.teams, '12')) {
            delete this.teams[12]
        }
        delete this.slots
        delete this.playerList
        delete this.buffer
        delete this.decompressed
        delete this.gameMetaDataDecoded
        delete this.header.blocks
        delete this.apmTimeSeries
    }

    finalize (): ParserOutput {
        const settings = {
            referees: !!this.meta.referees,
            fixedTeams: !!this.meta.fixedTeams,
            fullSharedUnitControl: !!this.meta.fullSharedUnitControl,
            alwaysVisible: !!this.meta.alwaysVisible,
            hideTerrain: !!this.meta.hideTerrain,
            mapExplored: !!this.meta.mapExplored,
            teamsTogether: !!this.meta.teamsTogether,
            randomHero: !!this.meta.randomHero,
            randomRaces: !!this.meta.randomRaces,
            speed: this.meta.speed
        }

        const root = {
            id: this.id,
            gamename: this.meta.gameName,
            randomseed: this.meta.randomSeed,
            startSpots: this.meta.startSpotCount,
            observers: this.observers,
            players: Object.values(this.players).sort(sortPlayers),
            matchup: this.matchup,
            creator: this.meta.creator,
            type: this.gametype,
            chat: this.chatlog,
            apm: {
                trackingInterval: this.playerActionTrackInterval
            },
            map: {
                path: this.meta.mapName,
                file: convert.mapFilename(this.meta.mapName),
                checksum: this.meta.mapChecksum
            },
            version: convert.gameVersion(this.header.version),
            duration: this.header.replayLengthMS,
            expansion: this.header.gameIdentifier === 'PX3W',
            settings,
            parseTime: Math.round(performance.now() - this.parseStartTime)
        }
        return root
    }
}

export default W3GReplay
