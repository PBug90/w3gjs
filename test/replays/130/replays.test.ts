import W3GReplay from "../../../src/";
import path from "path";
import { readFileSync } from "fs";
import { Validator } from "jsonschema";
import schema from "../../schema.json";

const Parser = new W3GReplay();
it("parses a standard 1.30.2 replay properly", async () => {
  const test = await Parser.parse(path.resolve(__dirname, "standard_1302.w3g"));
  expect(test.version).toBe("1.30.2+");
  expect(test.matchup).toBe("NvU");
  expect(test.players.length).toBe(2);
});

it("parses a standard 1.30.3 replay properly", async () => {
  const test = await Parser.parse(path.resolve(__dirname, "standard_1303.w3g"));
  expect(test.version).toBe("1.30.2+");
  expect(test.players.length).toBe(2);
});

it("parses a standard 1.30.4 replay properly", async () => {
  const test = await Parser.parse(path.resolve(__dirname, "standard_1304.w3g"));
  expect(test.version).toBe("1.30.2+");
  expect(test.players.length).toBe(2);
});

it("parses a standard 1.30.4 replay properly as buffer", async () => {
  const buffer: Buffer = readFileSync(
    path.resolve(__dirname, "standard_1304.w3g"),
  );
  const test = await Parser.parse(buffer);
  expect(test.version).toBe("1.30.2+");
  expect(test.players.length).toBe(2);
});

it("parses a standard 1.30.4 2on2 replay properly", async () => {
  const test = await Parser.parse(
    path.resolve(__dirname, "standard_1304.2on2.w3g"),
  );
  expect(test.version).toBe("1.30.2+");
  expect(test.buildNumber).toBe(6061);
  expect(test.players.length).toBe(4);
  expect(test.players[2]).toMatchSnapshot();
});

it("parses a standard 1.30 replay properly", async () => {
  const test = await Parser.parse(path.resolve(__dirname, "standard_130.w3g"));
  expect(test.version).toBe("1.30");
  expect(test.matchup).toBe("NvU");
  expect(test.type).toBe("1on1");
  expect(test.players[0].name).toBe("sheik");
  expect(test.players[0].race).toBe("U");
  expect(test.players[0].raceDetected).toBe("U");
  expect(test.players[1].name).toBe("123456789012345");
  expect(test.players[1].race).toBe("N");
  expect(test.players[1].raceDetected).toBe("N");
  expect(test.players.length).toBe(2);
  expect(test.players[0].heroes[0]).toEqual(
    expect.objectContaining({ id: "Udea", level: 6 }),
  );
  expect(test.players[0].heroes[1]).toEqual(
    expect.objectContaining({ id: "Ulic", level: 6 }),
  );
  expect(test.players[0].heroes[2]).toEqual(
    expect.objectContaining({ id: "Udre", level: 3 }),
  );
  expect(test.map).toEqual({
    file: "(4)TwistedMeadows.w3x",
    checksum: "c3cae01d",
    checksumSha1: "23dc614cca6fd7ec232fbba4898d318a90b95bc6",
    path: "Maps\\FrozenThrone\\(4)TwistedMeadows.w3x",
  });
});

it("parsing result has the correct schema", async () => {
  const test = await Parser.parse(path.resolve(__dirname, "standard_130.w3g"));
  const validatorInstance = new Validator();
  validatorInstance.validate(test, schema, { throwError: true });
});

it("resets elapsedMS instance property to 0 before parsing another replay", async () => {
  await Parser.parse(path.resolve(__dirname, "standard_130.w3g"));
  const msElapsed = Parser.msElapsed;
  await Parser.parse(path.resolve(__dirname, "standard_130.w3g"));
  const msElapsedTwo = Parser.msElapsed;
  expect(msElapsed).toEqual(msElapsedTwo);
});
