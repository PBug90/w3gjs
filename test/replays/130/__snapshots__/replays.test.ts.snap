// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses a standard 1.30.4 2on2 replay properly 1`] = `
{
  "actions": {
    "ability": 163,
    "assigngroup": 60,
    "basic": 115,
    "buildtrain": 124,
    "esc": 9,
    "item": 2,
    "removeunit": 0,
    "rightclick": 1433,
    "select": 988,
    "selecthotkey": 752,
    "subgroup": 0,
    "timed": [
      144,
      211,
      282,
      187,
      203,
      181,
      171,
      218,
      208,
      187,
      178,
      176,
      213,
      221,
      232,
      184,
      180,
      193,
      119,
    ],
  },
  "apm": 198,
  "buildings": {
    "order": [
      {
        "id": "hbla",
        "ms": 33036,
      },
      {
        "id": "halt",
        "ms": 36623,
      },
      {
        "id": "hbar",
        "ms": 38699,
      },
      {
        "id": "hhou",
        "ms": 46647,
      },
      {
        "id": "hhou",
        "ms": 62724,
      },
      {
        "id": "hhou",
        "ms": 122755,
      },
      {
        "id": "hkee",
        "ms": 190494,
      },
      {
        "id": "hhou",
        "ms": 210930,
      },
      {
        "id": "hwtw",
        "ms": 216711,
      },
      {
        "id": "hatw",
        "ms": 248942,
      },
      {
        "id": "hhou",
        "ms": 253012,
      },
      {
        "id": "hhou",
        "ms": 254028,
      },
      {
        "id": "hvlt",
        "ms": 258542,
      },
      {
        "id": "hhou",
        "ms": 327627,
      },
      {
        "id": "hhou",
        "ms": 485876,
      },
      {
        "id": "htow",
        "ms": 504685,
      },
      {
        "id": "hlum",
        "ms": 662483,
      },
      {
        "id": "hhou",
        "ms": 673657,
      },
      {
        "id": "hatw",
        "ms": 933869,
      },
      {
        "id": "htow",
        "ms": 1065471,
      },
    ],
    "summary": {
      "halt": 1,
      "hatw": 2,
      "hbar": 1,
      "hbla": 1,
      "hhou": 9,
      "hkee": 1,
      "hlum": 1,
      "htow": 2,
      "hvlt": 1,
      "hwtw": 1,
    },
  },
  "color": "#ff0303",
  "groupHotkeys": {
    "0": {
      "assigned": 1,
      "used": 128,
    },
    "1": {
      "assigned": 47,
      "used": 283,
    },
    "2": {
      "assigned": 7,
      "used": 45,
    },
    "3": {
      "assigned": 0,
      "used": 0,
    },
    "4": {
      "assigned": 1,
      "used": 211,
    },
    "5": {
      "assigned": 0,
      "used": 0,
    },
    "6": {
      "assigned": 0,
      "used": 0,
    },
    "7": {
      "assigned": 0,
      "used": 0,
    },
    "8": {
      "assigned": 2,
      "used": 2,
    },
    "9": {
      "assigned": 2,
      "used": 83,
    },
  },
  "heroes": [
    {
      "abilities": {
        "AHad": 2,
        "AHhb": 2,
      },
      "abilityOrder": [
        {
          "time": 166550,
          "type": "ability",
          "value": "AHhb",
        },
        {
          "time": 226152,
          "type": "ability",
          "value": "AHad",
        },
        {
          "time": 298939,
          "type": "ability",
          "value": "AHhb",
        },
        {
          "time": 592029,
          "type": "ability",
          "value": "AHad",
        },
      ],
      "id": "Hpal",
      "level": 4,
      "retrainingHistory": [],
    },
    {
      "abilities": {
        "AHbn": 1,
        "AHdr": 2,
      },
      "abilityOrder": [
        {
          "time": 387930,
          "type": "ability",
          "value": "AHdr",
        },
        {
          "time": 652084,
          "type": "ability",
          "value": "AHbn",
        },
        {
          "time": 819587,
          "type": "ability",
          "value": "AHdr",
        },
      ],
      "id": "Hblm",
      "level": 3,
      "retrainingHistory": [],
    },
  ],
  "id": 5,
  "items": {
    "order": [
      {
        "id": "sreg",
        "ms": 327162,
      },
      {
        "id": "plcl",
        "ms": 341634,
      },
      {
        "id": "tsct",
        "ms": 342990,
      },
      {
        "id": "phea",
        "ms": 417767,
      },
      {
        "id": "plcl",
        "ms": 619906,
      },
      {
        "id": "sreg",
        "ms": 690935,
      },
      {
        "id": "plcl",
        "ms": 692398,
      },
      {
        "id": "stwp",
        "ms": 694301,
      },
      {
        "id": "sreg",
        "ms": 829648,
      },
      {
        "id": "phea",
        "ms": 833926,
      },
      {
        "id": "sreg",
        "ms": 909258,
      },
      {
        "id": "phea",
        "ms": 910479,
      },
      {
        "id": "plcl",
        "ms": 918180,
      },
      {
        "id": "tsct",
        "ms": 926555,
      },
      {
        "id": "sreg",
        "ms": 932210,
      },
      {
        "id": "stel",
        "ms": 992084,
      },
      {
        "id": "shea",
        "ms": 995541,
      },
    ],
    "summary": {
      "phea": 3,
      "plcl": 4,
      "shea": 1,
      "sreg": 5,
      "stel": 1,
      "stwp": 1,
      "tsct": 2,
    },
  },
  "name": "Thorzain",
  "race": "H",
  "raceDetected": "H",
  "resourceTransfers": [
    {
      "gold": 0,
      "lumber": 30,
      "msElapsed": 82662,
      "playerId": 9,
      "playerName": "Starshaped",
      "slot": 3,
    },
    {
      "gold": 200,
      "lumber": 0,
      "msElapsed": 736480,
      "playerId": 9,
      "playerName": "Starshaped",
      "slot": 3,
    },
  ],
  "teamid": 3,
  "units": {
    "order": [
      {
        "id": "hpea",
        "ms": 812,
      },
      {
        "id": "hpea",
        "ms": 2885,
      },
      {
        "id": "hpea",
        "ms": 8742,
      },
      {
        "id": "hpea",
        "ms": 39120,
      },
      {
        "id": "hpea",
        "ms": 58282,
      },
      {
        "id": "hpea",
        "ms": 68247,
      },
      {
        "id": "hpea",
        "ms": 86734,
      },
      {
        "id": "hpea",
        "ms": 102922,
      },
      {
        "id": "hrif",
        "ms": 112469,
      },
      {
        "id": "hrif",
        "ms": 134194,
      },
      {
        "id": "hrif",
        "ms": 160527,
      },
      {
        "id": "hrif",
        "ms": 202902,
      },
      {
        "id": "hrif",
        "ms": 250426,
      },
      {
        "id": "hrif",
        "ms": 277323,
      },
      {
        "id": "hrif",
        "ms": 307310,
      },
      {
        "id": "hrif",
        "ms": 363259,
      },
      {
        "id": "hrif",
        "ms": 397380,
      },
      {
        "id": "hrif",
        "ms": 410317,
      },
      {
        "id": "hrif",
        "ms": 445278,
      },
      {
        "id": "hrif",
        "ms": 465282,
      },
      {
        "id": "hrif",
        "ms": 525932,
      },
      {
        "id": "hrif",
        "ms": 542208,
      },
      {
        "id": "hrif",
        "ms": 549121,
      },
      {
        "id": "hpea",
        "ms": 550446,
      },
      {
        "id": "hpea",
        "ms": 550581,
      },
      {
        "id": "hrif",
        "ms": 579797,
      },
      {
        "id": "hrif",
        "ms": 593823,
      },
      {
        "id": "hrif",
        "ms": 610863,
      },
      {
        "id": "hpea",
        "ms": 630067,
      },
      {
        "id": "hpea",
        "ms": 630179,
      },
      {
        "id": "hrif",
        "ms": 636477,
      },
      {
        "id": "hpea",
        "ms": 661473,
      },
      {
        "id": "hrif",
        "ms": 680850,
      },
      {
        "id": "hpea",
        "ms": 681776,
      },
      {
        "id": "hpea",
        "ms": 681972,
      },
      {
        "id": "hpea",
        "ms": 712775,
      },
      {
        "id": "hrif",
        "ms": 721842,
      },
      {
        "id": "hpea",
        "ms": 743320,
      },
      {
        "id": "hrif",
        "ms": 757033,
      },
      {
        "id": "hrif",
        "ms": 771398,
      },
      {
        "id": "hpea",
        "ms": 779182,
      },
      {
        "id": "hrif",
        "ms": 802580,
      },
      {
        "id": "hrif",
        "ms": 814565,
      },
      {
        "id": "hrif",
        "ms": 830732,
      },
      {
        "id": "hrif",
        "ms": 852290,
      },
      {
        "id": "hpea",
        "ms": 853277,
      },
      {
        "id": "hrif",
        "ms": 860086,
      },
      {
        "id": "hrif",
        "ms": 878292,
      },
      {
        "id": "hpea",
        "ms": 879865,
      },
      {
        "id": "hrif",
        "ms": 893893,
      },
      {
        "id": "hrif",
        "ms": 916553,
      },
      {
        "id": "hrif",
        "ms": 959420,
      },
      {
        "id": "hrif",
        "ms": 984615,
      },
      {
        "id": "hrif",
        "ms": 1006417,
      },
      {
        "id": "hrif",
        "ms": 1035124,
      },
      {
        "id": "hrif",
        "ms": 1045427,
      },
      {
        "id": "hrif",
        "ms": 1082176,
      },
      {
        "id": "hrif",
        "ms": 1112217,
      },
    ],
    "summary": {
      "hpea": 20,
      "hrif": 38,
    },
  },
  "upgrades": {
    "order": [
      {
        "id": "Rhri",
        "ms": 332547,
      },
      {
        "id": "Rhra",
        "ms": 623973,
      },
    ],
    "summary": {
      "Rhra": 1,
      "Rhri": 1,
    },
  },
}
`;
